### PR TITLE
rust/sip: register parser for tcp v9

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3351

Describe changes:
- Documentation updated
- Implements pattern matching for protocol detection
- Renames `append_request` and `append_response` to `build_tx_request` and `build_tx_response`
- Removes `input_len >= 3` condition
- SIP version characters are enforced to strictly match `SIP/2.0`

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_REPO=
SV_BRANCH=pr/1494
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

**Comment by catenacyber** [comment]
> SIP version characters are enforced to strictly match SIP/2.0

Why so ?
I think we want to match on past and future versions as well...

File: rust/src/sip/parser.rs [hunk 1]
```diff
@@ -78,7 +78,7 @@ fn is_request_uri_char(b: u8) -> bool {
 
 #[inline]
 fn is_version_char(b: u8) -> bool {
-    is_alphanumeric(b) || b"./".contains(&b)
+    b"SIP/2.0".contains(&b)
```
Comment (catenacyber) [inline rust/src/sip/parser.rs:81]: I think we rather want the caller of `is_version_char` begin by `tag("SIP/")`, what do you think ?

File: rust/src/sip/sip.rs [hunk 2]
```diff
@@ -518,6 +518,59 @@ pub unsafe extern "C" fn rs_sip_parse_response_tcp(
     state.parse_response_tcp(flow, stream_slice)
 }
 
+fn register_pattern_probe(proto: u8) -> i8 {
+    let methods: Vec<&str> = vec![
+        "REGISTER\0",
```
Comment (catenacyber) [inline rust/src/sip/sip.rs:523]: I do not think you need the final 0 in rust, do you ?

File: rust/src/sip/sip.rs [hunk 3]
```diff
@@ -518,6 +518,59 @@ pub unsafe extern "C" fn rs_sip_parse_response_tcp(
...
+        "PUBLISH\0",
+        "MESSAGE\0",
+        "INFO\0",
+        "OPTIONS\0",
+    ];
+    let mut r = 0;
+    unsafe {
+        for method in methods {
+            let depth = (method.len() - 1) as u16;
+            r |= AppLayerProtoDetectPMRegisterPatternCSwPP(
```
Comment (catenacyber) [inline rust/src/sip/sip.rs:542]: do we really need probing after this pattern ? 

Plus this commit should reference https://redmine.openinfosecfoundation.org/issues/5047 right ?

File: rust/src/sip/parser.rs (hunk 1 above)
Comment (glongo) [reply rust/src/sip/parser.rs:81]: I agree with you, it's much better tagging `SIP/` string first and then taking the chars until `\r\n` is found.

**Comment by glongo** [comment]
> > SIP version characters are enforced to strictly match SIP/2.0
> 
> Why so ? I think we want to match on past and future versions as well...

IIRC `SIP/2.0` is the only version so far, but it's better to tag and parse it as we already said

File: rust/src/sip/sip.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/sip/sip.rs:542]: So, I think it is rather `AppLayerProtoDetectPMRegisterPatternCS` to be used

File: rust/src/sip/sip.rs (hunk 2 above)
Comment (glongo) [reply rust/src/sip/sip.rs:523]: It seems that the same logic is applied in `bittorrent_dht.rs`

```
const BITTORRENT_DHT_PAYLOAD_PREFIX: &[u8] = b"d1:ad2:id20:\0";

if AppLayerProtoDetectPMRegisterPatternCS(
    IPPROTO_UDP,
    ALPROTO_BITTORRENT_DHT,
    BITTORRENT_DHT_PAYLOAD_PREFIX.as_ptr() as *const c_char,
    BITTORRENT_DHT_PAYLOAD_PREFIX.len() as u16 - 1,
    0,
    crate::core::Direction::ToServer.into(),
) < 0
{
    SCLogDebug!("Failed to register protocol detection pattern for direction TOSERVER");
};
```

Comment (catenacyber) [reply rust/src/sip/sip.rs:523]: https://redmine.openinfosecfoundation.org/issues/6583 ;-)

**Comment by glongo** [comment]
Replaced by https://github.com/OISF/suricata/pull/9909

**Pull request closed without merging**